java/src/main/native/src/row_conversion.cu

```diff
@@ -304,20 +334,600 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
+ *        this limit.
+ *
+ */
+struct row_batch {
+  size_type num_bytes;                   // number of bytes in this batch
+  size_type row_count;                   // number of rows in the batch
+  device_uvector<size_type> row_offsets; // offsets column of output cudf column
+};
+
+/**
+ * @brief Holds information about the batches of data to be processed
+ *
+ */
+struct batch_data {
+  device_uvector<size_type> batch_row_offsets; // offset column of returned cudf column
+  std::vector<size_type>
+      batch_row_boundaries;           // row numbers for different batches like 0, 10000, 20000
+  std::vector<row_batch> row_batches; // information about each batch such as byte count
+};
+
+/**
+ * @brief copy data from cudf columns into JCUDF format, which is row-based
+ *
+ * @param num_rows total number of rows in the table
+ * @param num_columns total number of columns in the table
+ * @param shmem_used_per_block shared memory amount each `block_info` is using
+ * @param block_infos span of `block_info` structs the define the work
+ * @param input_data pointer to raw table data
+ * @param col_sizes array of sizes for each element in a column - one per column
+ * @param col_offsets offset into input data row for each column's start
+ * @param row_offsets offset to a specific row in the output data
+ * @param output_data pointer to output data
+ *
+ */
+__global__ void copy_to_rows(const size_type num_rows, const size_type num_columns,
+                             const size_type shmem_used_per_block,
+                             device_span<const block_info> block_infos, const int8_t **input_data,
+                             const size_type *col_sizes, const size_type *col_offsets,
+                             const size_type *row_offsets, int8_t **output_data) {
+  // We are going to copy the data in two passes.
+  // The first pass copies a chunk of data into shared memory.
+  // The second pass copies that chunk from shared memory out to the final location.
+
+  // Because shared memory is limited we copy a subset of the rows at a time.
+  // This has been broken up for us in the block_info struct, so we don't have
+  // any calculation to do here, but it is important to note.
+
+  constexpr unsigned stages_count = NUM_BLOCKS_PER_KERNEL_LOADED;
+  auto group = cooperative_groups::this_thread_block();
+  extern __shared__ int8_t shared_data[];
+  int8_t *shared[stages_count] = {shared_data, shared_data + shmem_used_per_block};
+
+  __shared__ cuda::barrier<cuda::thread_scope_block> block_barrier[NUM_BLOCKS_PER_KERNEL_LOADED];
+  if (group.thread_rank() == 0) {
+    for (int i = 0; i < NUM_BLOCKS_PER_KERNEL_LOADED; ++i) {
+      init(&block_barrier[i], group.size());
+    }
+  }
+
+  group.sync();
+
+  auto const blocks_remaining =
+      std::min(static_cast<uint>(block_infos.size()) - blockIdx.x * NUM_BLOCKS_PER_KERNEL_TO_ROWS,
+               static_cast<uint>(NUM_BLOCKS_PER_KERNEL_TO_ROWS));
+
+  size_t fetch;
```

Comment:
I'm having a little trouble keeping these straight in my head.
`fetch` refers to the stage of the copy? And `subset` refers to the tile(s) that the current GPU block is processing?

Reply:
Correct. I didn't choose these and left them to match the documentation for `memcpy_async`, but I think some more useful names are in order and possibly even less clever filling of the buffer.

Reply:
Changed the names and commented them, see if this works out nicer for you.